### ISSUE TITLE
Incorrect history in 03_gradient/plot_loss_with_path.py?

### ISSUE THREAD

**Issue by skoe** [nusco/progml-code#1]
First of all, thanks for the great book and for keeping the code up to date! There seems to be a bug in 03_gradient/plot_loss_with_path.py, if I understood everything correctly.

The history path should be on the surface, because the gradient descent moves along the surface, right? When I rotate the diagram, the line floats partly over the surface and has two direction changes. 

![Figure_1](https://github.com/user-attachments/assets/9eea9274-8c0d-44f2-9f54-09fa7b9630b5)

I couldn't understand that. If I change the code like this:

```patch
         current_loss = loss(X, Y, w, b)
-        history.append([w, b, previous_loss])
+        history.append([w, b, current_loss])
```

![Figure_2](https://github.com/user-attachments/assets/d61eb3e0-172e-48e1-887f-aaba2c3ccaeb)

Now the line lies on the surface, it only has one kink and now actually seems to follow the steepest descent in each case. It also makes sense according to my understanding, because we want to plot the loss that belongs to w, b not the loss that belonged to the previous w and previous b. Is that correct? If so, perhaps you can put this corrections (and others, if any) in a README in the appropriate directory?

**Comment by nusco** [comment]
Hey, @skoe! Dang, you're right! I update the weight and bias, but then I plot the loss _before_ the parameter update. I didn't even notice the glitch in the chart, because I thought that the GD path was following the surface in the direction of a higher bias and weight. Instead, the plotted segment is floating over the surface, as you say. Very well spotted!

I won't fix this code straight away, because I want this repo to be consistent with the book's code. However, I added a note to my todo for a second edition. That one will take a while, because I'm busy with another book–but it will probably happen at some point. Thank you!
